Grading Table

Toemmsche opened the pull request:
# Added
- Interactive grading table that is displayed in the submission overview
- local and remote grading table persistence

# Changed
- gradingTable is now a required property of Exam

Toemmsche committed:
WIP

Toemmsche committed:
[+] Added grading table with UI, local and remote persistence
[~] gradingTable required property on Exam

Toemmsche committed:
[~] Sort grading table lower bounds

Toemmsche committed:
Merge branch 'dev' into feature/grading-table

# Conflicts:
#	src/lib/blocs/remark/remark_bloc.dart
#	src/lib/blocs/remark/remark_state.dart
#	src/lib/components/correction/correction_overview.dart
#	src/packages/schoolexam/lib/exams/dto/exam_dto.dart
#	src/packages/schoolexam/lib/exams/hybrid_exams_repository.dart
#	src/packages/schoolexam/lib/exams/local_exams_repository.dart
#	src/packages/schoolexam/lib/exams/online_exams_repository.dart

Toemmsche committed:
[+] Re-Add local persistence for grading table after merge

Toemmsche committed:
[+] Added grading table localization

Toemmsche committed:
[+] Add 0 to 15 grading scheme (oberstufe)
[+] Add dropdown button for grading scheme selection

Toemmsche committed:
[+] Add "official" grading schemes (source: wikipedia.org)

Toemmsche committed:
[~] Change grade descriptor for 4-

Toemmsche committed:
[~] Use half points for default grading schemes
[~] Optimize imports

tosoikea commented on src/packages/schoolexam/lib/exams/models/grading_table_lower_bound.dart:
```diff
@@ -0,0 +1,34 @@
+
+  static const empty = GradingTableLowerBound(points: 0.0, grade: "");
+
+  GradingTableLowerBoundDTO toDTO() {
```
Mir würde es persönlich besser gefallen, wenn die DTO eine fromModel definiert und die Umwandlung nicht im Model drin ist.

Toemmsche replied:
Fixed.

tosoikea commented on src/packages/schoolexam/lib/exams/models/grading_table.dart:
```diff
@@ -0,0 +1,25 @@
+        lowerBounds: lowerBounds.map((lb) => lb.valueCopy()).toList());
+  }
+
+  GradingTableDTO toDTO() {
```
Mir würde es persönlich besser gefallen, wenn die DTO eine fromModel definiert und die Umwandlung nicht im Model drin ist.

Toemmsche replied:
Fixed.

tosoikea commented on src/packages/schoolexam/lib/exams/dto/exam_dto.dart:
```diff
@@ -40,11 +43,13 @@ class ExamDTO extends Equatable {
             .map((e) => TaskDTO.fromJson(e))
-            .toList();
+            .toList(),
+        gradingTable = json['gradingTable'] != null ? GradingTableDTO.fromJson(json['gradingTable']) : null;
```
Das ist eine Map, dann lieber containsKey()

Toemmsche replied:
`"gradingTable"` ist immer als key vorhanden, aber z.T. null. `GradingTableDTO.fromJSON(json['gradingTable'])` kann aber nicht mit null Werten umgehen.

tosoikea commented on src/lib/components/correction/correction_overview.dart:
```diff
@@ -11,10 +12,15 @@ class CorrectionOverview extends StatelessWidget {
+      builder: (context, state) => Column(
+            children: [
+              const CorrectionGradingTable(),
+              Expanded(
```
Expanded macht hier irgendwie nur sehr eingeschränkt Sinn.

Toemmsche replied:
Ohne Expanded entsteht ein RenderFlexOverflow. Siehe https://stackoverflow.com/q/49480051

tosoikea commented on src/lib/blocs/remark/remark_bloc.dart:
```diff
@@ -279,6 +284,143 @@ class RemarkCubit extends Cubit<RemarkState> {
   }
 
+  /// Create an empty grading table
+  void initGradingTable() {
```
Welche Funktionalität ist hier gegeben?

Toemmsche replied:
Artefakt aus Zeiten, in denen `gradingTable` noch kein Attribut von `Exam` war. Wird entfernt.

tosoikea commented on src/lib/blocs/remark/remark_bloc.dart:
```diff
@@ -279,6 +284,143 @@ class RemarkCubit extends Cubit<RemarkState> {
+  /// Change the grading table to the default layout
+  /// The two standard german grading schemes are available as presets
+  /// Source: https://de.wikipedia.org/wiki/Vorlage:Punktesystem_der_gymnasialen_Oberstufe
+  void getDefaultGradingTable(int low, int high) {
```
Das hier eventuell in eine Helper-Klasse, sonst blähen wir weiter den RemarkCubit auf.

Toemmsche replied:
Fixed.

tosoikea commented on src/lib/components/correction/correction_grading_table.dart:
```diff
@@ -0,0 +1,250 @@
+  }
+
+// Helper function to show a single highlighted widget in a modal bottom sheet
+  _showSingleWidgetBottomSheet(
```
Type?

Toemmsche replied:
`VoidCallBack`. Fixed.

tosoikea replied:
So ist die Funktion vom Typ dynamic. Müsste void _showSingleWidgetBottomSheet sein. Spielt für die Funktionalität aber keine Rolle.

Toemmsche replied:
Die Funktion gibt eine Funktion zurück.

tosoikea commented on src/lib/components/correction/correction_grading_table.dart:
```diff
@@ -0,0 +1,250 @@
+import 'package:modal_bottom_sheet/modal_bottom_sheet.dart';
+import 'package:schoolexam_correction_ui/blocs/remark/remark.dart';
+
+/// This widget allows for editing of the grading table that should be assigned to the exam
```
Das schon ein Bomben-Widget, uff. Kann man das evtl. in etwas kleinere Pakete unterteilen?

Toemmsche replied:
Habe den Header und die Action Bar extrahiert.

Toemmsche committed:
[~] replaced toDTO() with fromModel()
[~] Broke up grading table widget into subclasses
[~] Moved default grading scheme logic into helper class
[~] Adjusted localization

tosoikea commented on src/packages/schoolexam/lib/exams/models/grading_table_lower_bound.dart:
```diff
@@ -0,0 +1,29 @@
+  }
+
+  /// Creates a copy by value of this grading table lower bound
+  GradingTableLowerBound valueCopy() {
```
Redundant zu copyWith ohne Parameter. Ist aber ok.
